proposals/break-continue-in-inline-lambdas.md

```diff
@@ -0,0 +1,175 @@
+    }
+}
+```
+
+The alternative approach may be to make `break`/`continue` apply to the closest `for`, `while` or "loop-like" functions. We will
+never allow unlabeled `break`/`continue` for "loop-like" functions. The rationale is simple: if Kotlin allowed unlabeled
+`break`/`continue` for "loop-like" functions then it would mean that for such code
+
+```kotlin
+while (cond) {
+    someInlineFunctionThatAcceptsLambda {
+        break
+    }
+}
+```
+
+users wouldn't be able to understand whether `break` breaks from `someInlineFunctionThatAcceptsLambda` or from `while` without
+looking into `someInlineFunctionThatAcceptsLambda` implementation. It complicates reasoning about `break` and `continue`
+statements. In other words, it's hard to distinguish "loop-like" functions from other lambdas if the call-site doesn't differ
+syntactically.
+
+Also, right now we have such an invariant in Kotlin: a pair of **keywords** `break`/`continue` tied to the closest `while` or
+`for` **keywords**. `return` **keyword** is tied to the closest `fun` **keyword**. The proposed alternative solution breaks this
+invariant.
+
+* **Q**: Why not make labels mandatory for non-local `break`/`continue`? That would eliminate `forEach` puzzler.
+* **A**: Because Kotlin doesn't require labels for non-local `return`. For sake of consistency, We should either require labels
+  for both `break`/`continue` and `return` cases or don't require the labels again for the both cases.
+
+* **Q**: Ok, why not make labels mandatory for both non-local `return` and non-local `break`/`continue` then?
+* **A**: We can't make non-local `return` to be obligatory labeled because it would break inline functions first-class
+  citizenship in language constructs. E.g. we value "return if not null" idiomatic Kotlin construct `nullable?.let { return it }`.
+  If non-local `return` were obligatory labeled then the code would be much more verbose
+  `nullable?.let { return@functionName it }`.
+
+It worth noticing that we don't abandon the idea to provide some "break"/"continue" mechanism for "loop-like" functions
+[KT-19748](https://youtrack.jetbrains.com/issue/KT-19748). It just that we agree that we will never support unlabeled non-local
+`break`/`continue` for "loop-like" functions.
+
+## Tooling for puzzling code
+
+Some users may find constructions containing `break` or `continue` within lambdas passed to "loop-like" inline functions
+such as `forEach` or `takeWhile` puzzling:
+```kotlin
+fun printUntilZero(producer: () -> List<Int>) {
+    while(true) {
+        val list = producer()
+        list.forEach {
+            if (it == 0) break // Does it exit from while or from forEach?
+            println(it)
+        }
+    }
+}
+```
+
+To reduce the risk of confusion, it was proposed to introduce an IDE inspection and a quickfix recommending to use labeled `break`
+and `continue` in lambdas passed to parameters without an `EXACTLY_ONCE` or `AT_MOST_ONCE` contract (potentially "loop-like").
```

Comment:
> recommending to use labeled `break` and `continue` in lambdas passed to parameters without an `EXACTLY_ONCE` or `AT_MOST_ONCE` contract

Concern: it means that we will warn users about non-local break and continue for user-defined inline functions because usually, users don't define contracts for own functions. Contracts are still experimental and we can't afford requiring users to use OptIn annotations

Reply:
Right now I'm thinking that the inspection scope should be restricted to checking only Stdlib functions

Reply:
@nikitabobko That's a valid concern. Not sure if limiting the inspection scope to Stdlib is a good idea. If we cannot make it work consistently, maybe we shouldn't implement it at all.

Reply:
Tying the inspection to the contract is a good option. 

It will typically be used only with stdlib, and we must be sure it is not triggered on the usual scope functions. It is not bad if it is triggered on a user-defined function without a contract. It is better to be safe (by suggesting to use labelled break/continue) than end up with a more ambiguous code on a user-defined loop-like function.

Reply:
Ok, since it's an inspection for an unreleased language feature we can try to be more aggressive because we don't color a lot of existing code in yellow. Users will file a ticket if it turns out that the inspection is too aggressive